[backport 5.1] Update zoekt to include updated grpc prometheus 2.0.0 version (#56735)

ggilmore opened the pull request:
This is a manual cherry pick of https://github.com/sourcegraph/sourcegraph/pull/56735, since there were merge conflicts. 

In addition, this PR sets an additional environment variable in the zoekt-sourcegraph-indexserver PR that ensures that grpc is off by default in the 5.1.X releases. See https://github.com/sourcegraph/sourcegraph/pull/56740/commits/b565eada481f487c0c8e1d9f2a248219c9d310d3 for more information.

 Conflicts:
	- deps.bzl
	- go.mod
	- go.sum



## Test plan

CI

burmudar commented on deps.bzl:
```diff
@@ -3399,19 +3431,19 @@ def go_dependencies():
-        version = "v2.0.0-rc.3",
+        importpath = "github.com/grpc-ecosystem/go-grpc-middleware/providers/prometheus",
+        sum = "h1:mdLirNAJBxnGgyB6pjZLcs6ue/6eZGBui6gXspfq4ks=",
+        version = "v1.0.0-rc.0",
```
Is this downgrade intentional?

ggilmore replied:
This is intentional. 

We were using a pre-release version of https://github.com/grpc-ecosystem/go-grpc-middleware. This is a repository that contains multiple sub-packages.

At https://github.com/grpc-ecosystem/go-grpc-middleware/tree/v2.0.0-rc.3, the package was named https://github.com/grpc-ecosystem/go-grpc-middleware/tree/v2.0.0-rc.3/providers/opentracing. 

At https://github.com/grpc-ecosystem/go-grpc-middleware/tree/v2.0.0, the package was renamed to https://github.com/grpc-ecosystem/go-grpc-middleware/tree/v2.0.0/providers/prometheus, and now has a 1.0.0 version (that you can see here in this example: https://github.com/grpc-ecosystem/go-grpc-middleware/blob/972f219a4c56d136ead1a9d1b823147d74d0bfae/examples/go.mod#L7)

burmudar commented:
Hey @ggilmore I updated the PR to make backcompat pass. You were on the right track!

1. We have to update the find sed expressions in `backcompat/defs.bzl`, so that we replace all repository definitions with the  _backcompat_ go repo definition in the v5.0.0 of the code. Otherwise it keep on using the original go repo definition. This is why the repo definition is prefixed with `back_compat_` in backcompat.
2. Once we've replaced all the repo definitions, we also have to tell _any_ dependencies that they must resolve _their_ dependencies to the backcompat go repo definition. This is done with a `gazelle:resolve` build directive. This took some trial and error to figure out 😮‍💨

ggilmore commented:
> Hey @ggilmore I updated the PR to make backcompat pass. You were on the right track!
> 
>     1. We have to update the find sed expressions in `backcompat/defs.bzl`, so that we replace all repository definitions with the  _backcompat_ go repo definition in the v5.0.0 of the code. Otherwise it keep on using the original go repo definition. This is why the repo definition is prefixed with `back_compat_` in backcompat.
> 
>     2. Once we've replaced all the repo definitions, we also have to tell _any_ dependencies that they must resolve _their_ dependencies to the backcompat go repo definition. This is done with a `gazelle:resolve` build directive. This took some trial and error to figure out 😮‍💨

Got it, thanks so much!

ggilmore commented:
@sourcegraph/release-guild Can someone merge this in? Thanks!
